### ISSUE TITLE
RFE - Implement fact cache filtering in job template options

### ISSUE THREAD

**Issue by zaskan** [ansible/awx#9403]
CUSTOMER CASE

We experience regular performance issues in form of out-of-memory situations caused by use of fact cache on large jobs (>100 endpoints), which we can easily reproduce in an isolated environment.

**Steps to reproduce:**

1. Create an inventory from the linked file [inventory.ini.txt](https://github.com/ansible/awx/files/6035570/inventory.ini.txt)
2. Create a job template using the playbook [test-env.yml.txt](https://github.com/ansible/awx/files/6035594/test-env.yml.txt)
 and enable `Use fact cache` option. JT should run against the inventory defined in (1) without defined limit (on all endpoints). Run this JT only once to populate fact for each host in inventory (you will notice the problem at this step as well)
3. Create a job template using the playbook [test.yml.txt](https://github.com/ansible/awx/files/6035600/test.yml.txt) and enable `Use fact cache` option. JT should run against the inventory defined in (1) without defined limit (on all endpoints). Run this JT as many times as you want to reproduce the issue.

**Observed behavior:**

- Increased memory usage 

![issue](https://user-images.githubusercontent.com/24324312/108994607-4df7e700-769c-11eb-9dc5-87c09182ec11.png)

- Increased storage usage due to Ansible Tower off-loading fact cache for all hosts in inventory (that's not exactly an issue, but is something we noticed)

**Workarounds:**

- Disable `Use fact cache` in job templates 

> While this is an option, it begs the question why Tower has support for fact cache at all. It also puts strain on the hosts, since playbook developers would have to collect required. It may also require re-design of playbooks and move of some key facts into vars.

**How it could be improved**
Following is based on our experiments with Ansible AWX project (we are able to reproduce the problem with it as well). 
As mentioned above, Ansible Tower off-loads entire inventory in `/tmp/awx`_... folder created for each job. Fact cache part of inventory is ending in `artifacts` folder. Ansible then uses those facts (by loading all of them in memory) during `ansible-playbook` command. 
We look at Ansible AWX code, and found that 

1. Ansible Tower has no functionality to filter fact cache before job launch. If user knows that the playbook selected for Job Template uses only 2 facts: `ansible_distribution` and `ansible_distribution_major_version`, then there is no reason  to feed entire set of facts to the job. Of course the default will stay as is (all facts would be used), but this RFE would give user an option to select which fact apply to the job. 
2. Fact cache manipulation is basically a `rip and replace` operation. 

Easiest way to "fix" this behavior, was to modify code for launching jobs to mirror functionality Ansible has in `setup` module, specifically implementing a way to apply a filter. Since we din't want to alter UI/API/DB model, this is done by checking existence of a specific variable in JT `extra_vars` 

![job-details](https://user-images.githubusercontent.com/24324312/108994084-a4b0f100-769b-11eb-8d2d-8d6da0a0f7fd.png)

You can find a run-time comparison with the fix and without the fix below.

![job-comparison](https://user-images.githubusercontent.com/24324312/108994931-af1fba80-769c-11eb-9a5b-1ae4a3afb294.png)

We also include the `.patch` file which can be applied to `devel` branch in `ansible/awx` project. The patch adds support for fact cache filtering through the use of `extra_vars`, ensures merging of `fact_cache` back into Tower inventory, unit tests.

[devel...egmar_fact-cache-pre-launch-filter.patch.txt](https://github.com/ansible/awx/files/6035612/devel.egmar_fact-cache-pre-launch-filter.patch.txt)

**Additional information**
As mentioned in step 2 of `steps to reproduce` you don't even need to have fact cache enabled. It's quite easy to achieve an out-of-memory scenario, just by having Ansible playbook generate a lot facts (which obviously are stored in memory). Though, as mentioned, this is not a major problem since one can limit what Ansible collects using `setup` module.

We already tried to set `gather_subset` and `ANSIBLE_INJECT_FACT_VARS=False` already. Of course, hard to say, if in all cases, as I'm sure various users might have a need to collect all facts, or even some custom facts they define on their systems.

**Comment by wenottingham** [comment]
The recommendation is to filter the facts on import, by using `gather_subset`; this is under the playbook user's control already, and they should be able to therfore use what is the minimum they need.

Using extra_vars makes the playbook operate differently inside and outside of the platform, especially if varialbes are remapped as is described above.

**Comment by egmar** [comment]
Hi Bill,

we know about `gather_subset` and we use it, though there are cases when users want to collect everything (can't stop them doing this). Some even make their custom facts and telling them not to use facts would be rather strange.

The main idea here was to have a similar mechanism, but as part of `JobOptions`. Currently only `use_fact_cache` exists, and it would be nice to have an additional option to specify a list of facts or a subset filter similar to one used by `setup` module, so only a subset of facts would be passed to the job.

The prototype included in issue was done using `extra_vars` just for the sake of simplicity, and as a proof that this idea is viable. A final version, if implemented, would definitely have no place in `extra_vars`, and rather be a separate field, perhaps with an advanced UI.

Nevertheless, for us the way Tower/AWX behaves at the moment, is kind of problematic. Having jobs consuming ~20GB of memory just for sake of 4 facts is strange, and at the same time, we would still like to allow users to be able to cache facts they collect or declare in their playbooks.

Ultimately, the problem it's not only in Tower/AWX, but spans into Ansible engine to some extent. See https://github.com/ansible/ansible/issues/73654
But even if it is fixed in Ansible engine, there is still the problem of unnecessary storage consumption caused by `use_fact_cache`. A job for 2000 hosts with 2 MB of facts per host, is basically 4 GB. Not to mention the delay in job launch caused by time used for writing those facts to disk.

**Comment by wenottingham** [comment]
`extra_vars` vs a specific model field isn't the distinction here - it's still a user configuration item that causes the playbook to run differetly inside and outside of AWX, in a way that can't be replicated outside (as opposed to, say, passing a limit).

I'd be against solving this in AWX just because of the coordination that's required - for each template, the template author would need to explicitly work with the content author as to the set of facts needed, update it across different versions of the playbook, and so on. By constraining this to be solved at the playbook level with more explicit fact collection, it resolves that need for coordination of values in two places.

**Comment by egmar** [comment]
@wenottingham it can be replicated outside of AWX by using same method proposed in RFE. e.g. `jq`/`yq` to filter into a temporary location, run playbook with custom `ansible.cfg`, re-merge facts back into origin again with `jq`/`yq`.

In our case, developers are talking to users and vice-versa, or even better have automations which pre-configure job templates as required, as such I do not see this as a huge problem. 

Our concern is not in collection of facts. Ansible gives us sufficient tools to control that, and if someone wants to collect a lot of facts, that can be done in controlled way. Our problem, is in the facts which already exist in Tower/AWX, they can be quite big (~600kb is quite common), and as I mentioned already, it would be strange to tell users not to use them. As an example, we have users collecting data from `ohai` + `ansible_facts` and then use this in host filter for smart inventories. Telling them they wont have this anymore, might not be an option.

The ideal solution could be a fix for https://github.com/ansible/ansible/issues/73654, but that's not likely to happen in near-term without a complete rewrite of how Ansible engine handles facts, hostvars and other type of variables, at least from my code analysis.
